### PR TITLE
Implemented getNumPassesToRobot evaluation function

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Implemented the `getNumPassesToRobot` evaluation function that will be used in the `calculateEnemyThreat` functions for #477. This function turned out to be quite a doozy so I've put it in a separate PR.

### Testing Done

Added unit tests

### Resolved Issues

None (but will be part of #477)

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: I'm a bit confused about what this means. Are we just removing the `initial_passer`?

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: since this has a fairly descriptive name, could we put it in the `robot.cpp` file? Seems like a fairly general thing that we will likely use again at some point, and it would be good to have it in an obvious place to avoid duplication.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: mind :boom:

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: Could we make this `for` loop into a function like `std::map<Robot, std::vector<Robot>, cmpRobotByID> findAllRecieverPasserPairs(std::vector<Robot> robots)`? Would probably want to move some of doc comments along with it.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: Could we make part or all of this `for` loop into a function like `std::vector<Robot> getAllPossibleReceivers(Robot passer, std::vector<Robot> potential_receivers)`. Would probably want to move some of doc comments along with it.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: Would `other_robots` ever be anything other then the enemy team?

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: I'm a bit confused about what this last sentence means?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Logic seems OK, but I struggled to read through it entirely just because the function is so long, so I'd like to take another look once we've worked through some of the comments I've left. Pretty much everything is focused on splitting the monolithic 150+ line function up. I think with the comments I've left we can get it down to ~90 with comments (~50 without), which is still larger then I'd like but should be much more readable. Note that all these new functions don't need to be public (they could just be declared and defined in the `.cpp`), but on the other hand it would be good to test them, so they'd need to be in the header for that. I'll leave that up to you.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: Yes that's all we're doing, since the algorithm already starts by "visiting" the initial passer. I've made this a bit more clear in the comments.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: Moved the comparator and updated the comments.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: _it's so pythonic_ :snake:

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: I don't think so. After some thought I'll change this to another `Team`.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: We already check the case where the passer and receiver are the same. If this was the case that would require `0` passes. So when we start the loop we are checking for the possibility of the receiver getting the ball in `1` pass. I'll try make the comments more clear.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: After some discussion on Slack I won't be doing separating this loop into a function.

File: src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp]: I separated this loop into its own function and added more unit tests for it.

**Review by garethellis0 — DISMISSED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
Fixed unit tests will need re-approval

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**
